### PR TITLE
move the engine download code to dart

### PR DESCRIPTION
**Pull request by devoncarew**
- move the engine download code to dart

`cache.dart` now manages `bin/cache/pkg`, `bin/cache/artifacts/engine`, and `bin/cache/artifacts/material_fonts`. Also, this PR starts downloading the `android-x64` engine artifacts.

### CODE REVIEW

File: packages/flutter_tools/lib/src/cache.dart [hunk 1]
```diff
@@ -50,6 +52,16 @@ class Cache {
...
 
+  Future<Null> updateAll() async {
+    MaterialFonts materialFonts = new MaterialFonts(cache);
+    if (!materialFonts.isUpToDate())
+      await materialFonts.download();
+
+    FlutterEngine engine = new FlutterEngine(cache);
+    if (!engine.isUpToDate())
+      await engine.download();
+  }
```
Comment (abarth) [inline packages/flutter_tools/lib/src/cache.dart]: Should we download these in parallel?

File: packages/flutter_tools/lib/src/cache.dart [hunk 2]
```diff
@@ -120,3 +132,91 @@ class MaterialFonts {
...
+      }
+    }
+
+    cache.setStampFor(kName, cache.getVersionFor(kName));
+  }
+
+  void _makeFilesExecutable(Directory dir) {
+    for (FileSystemEntity entity in dir.listSync()) {
+      // chmod a+x "$ENGINE_ARTIFACT_PATH/darwin-x64/sky_snapshot" and sky_shell
+      if (entity is File && path.basename(entity.path).startsWith('sky_'))
```
Comment (abarth) [inline packages/flutter_tools/lib/src/cache.dart]: This is sort of magical.  We'll probably rename these at some point.  I guess it was pretty magical before too.

**Comment by abarth** [comment]
LGTM but I'd probably structure the await calls so that all the downloads happen in parallel.

File: packages/flutter_tools/lib/src/cache.dart (hunk 2 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/cache.dart]: I changed this to looking for `sky_snapshot` or `sky_shell` explicitly. We could also choose to tar stuff, which will handle the permissions better.

File: packages/flutter_tools/lib/src/cache.dart (hunk 1 above)
Comment (devoncarew) [inline packages/flutter_tools/lib/src/cache.dart]: I'd prefer not to :grimacing: ; which might sound odd, but the progress display will work better if there's just one task on-going at a time, and the total download time w/ an engine rev is just ~10 seconds.

Comment (abarth) [inline packages/flutter_tools/lib/src/cache.dart]: Ok

**Comment by Hixie** [comment]
you could also just only show one progress bar but still run the download in the background. or show an aggregate progress bar.

**Pull request merged**